### PR TITLE
do-agent: init at 3.5.6

### PR DESCRIPTION
**Pull request by yvt**
###### Motivation for this change

This agent program is needed for DigitalOcean's [monitoring panel](https://www.digitalocean.com/docs/droplets/how-to/graphs/) and alert feature to work. It’s highly useful for those who use NixOS on DO droplets.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/servers/monitoring/do-agent/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{ lib, buildGoModule, fetchFromGitHub }:
+
+buildGoModule rec {
+  name = "do-agent-${version}";
```
Comment (alexarice) [inline pkgs/servers/monitoring/do-agent/default.nix]: please use pname

File: pkgs/servers/monitoring/do-agent/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  buildFlagsArray = ''
+    -ldflags=
+      -X main.version=${version}
+  '';
+
+  modSha256 = "0ydjwxdkcz0blpzwapiaq66vh7nrcg4j91z6h7v4ynnw2rgp7pmy";
+
+  meta = with lib; {
+    description = ''
```
Comment (alexarice) [inline pkgs/servers/monitoring/do-agent/default.nix]: description should only be one line long. You can include a longDescription if you feel it is necessary

**Comment by alexarice** [comment]
@GrahamcOfBorg build do-agent

**Comment by alexarice** [comment]
Should this work on aarch?

**Comment by yvt** [comment]
It should theoretically work on AArch64, too, though DO doesn't provide Arm-based droplets at this moment.

**Comment by alexarice** [comment]
@GrahamcOfBorg build do-agent

**Review by alexarice — APPROVED** [review]
Builds with nix-review though I can't really comment on the module

**Comment by Kiwi** [comment]
The module is working for me with defaults at least. Thanks!

File: maintainers/maintainer-list.nix [hunk 3]
```diff
@@ -7064,6 +7064,11 @@
     githubId = 1866448;
     name = "Eric Bailey";
   };
+  yvt = {
+    email = "i@yvt.jp";
+    github = "yvt";
```
Comment (Infinisil) [inline maintainers/maintainer-list.nix]: Needs `githubId` as well

File: nixos/modules/services/monitoring/do-agent.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  options.services.do-agent = {
+    enable = mkEnableOption "do-agent, the DigitalOcean droplet metrics agent";
+
+    user = mkOption {
+      type = types.str;
+      default = "do-agent";
+      description = "User account under which do-agent runs.";
+    };
+
+    group = mkOption {
```
Comment (Infinisil) [inline nixos/modules/services/monitoring/do-agent.nix]: Is there a need for user/group to be configurable? Also try to see if `DynamicUser` works for this.

File: nixos/modules/services/monitoring/do-agent.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.do-agent ];
+
+    systemd.services.do-agent = {
+      description = "DigitalOcean Droplet Metrics Agent";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      serviceConfig = {
```
Comment (Infinisil) [inline nixos/modules/services/monitoring/do-agent.nix]: You need to set `serviceConfig.User` for it to use the created user

File: nixos/modules/services/monitoring/do-agent.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.do-agent ];
+
+    systemd.services.do-agent = {
+      description = "DigitalOcean Droplet Metrics Agent";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
```
Comment (Infinisil) [inline nixos/modules/services/monitoring/do-agent.nix]: I think this should probably be `after = [ "network-online.target" ]` and `wants = [ "network-online.target" ]` as per `man systemd.special`

File: nixos/modules/services/monitoring/do-agent.nix [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        PrivateTmp = "yes";
+        ProtectSystem = "full";
+        ProtectHome = "yes";
+        NoNewPrivileges = "yes";
+      };
+    };
+
+    users.users = optionalAttrs (cfg.user == "do-agent") (singleton
+      { name = "do-agent";
+        group = cfg.group;
```
Comment (Infinisil) [inline nixos/modules/services/monitoring/do-agent.nix]: If `DynamicUser` doesn't work (but I really think it should, might need some extra permissions though), then you should set `isSystemUser = true`. See NixOS/rfcs#52 too

**Review by infinisil — CHANGES_REQUESTED** [review]

**Review by infinisil — APPROVED** [review]
Cool, tested it with my DO droplet, working well, thanks for the PR!

**Comment by Infinisil** [comment]
Ah whoops, should've squashed the commits a bit, oh well

**Pull request merged**